File: public/demos/topic_50752/data_analysis_1.1/server/llm.js
/**
 * OpenAI 兼容的 Chat Completions 客户端。
 * 支持流式输出与工具调用（function calling）增量解析。
 */

function normalizeBaseUrl(baseUrl) {
  let url = (baseUrl || "").trim().replace(/\/+$/, "");
  if (!url) url = "https://api.openai.com/v1";
  if (!/\/(v1|chat|completions|openai)/i.test(url) && /^https?:\/\/[^/]+$/.test(url)) {
    url += "/v1";
  }
  return url;
}

export async function streamChat(config, messages, tools, handlers = {}, signal = null) {
  const base = normalizeBaseUrl(config.baseUrl);
  const endpoint = `${base}/chat/completions`;

  const body = {
    model: config.model,
    messages,
    stream: true,
    temperature: config.temperature ?? 0.2,
  };
  if (tools && tools.length) {
    body.tools = tools;
    body.tool_choice = "auto";
  }

  const resp = await fetch(endpoint, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${config.apiKey || ""}`,
    },
    body: JSON.stringify(body),
    signal: signal ?? undefined,
  });

  if (!resp.ok || !resp.body) {
    const text = await resp.text().catch(() => "");
    throw new Error(`LLM 接口错误 ${resp.status}: ${text.slice(0, 500)}`);
  }

  const reader = resp.body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";

  let content = "";
  const toolCallsMap = new Map();
  let finishReason = null;

  const handleEvent = (dataStr) => {
    if (dataStr === "[DONE]") return;
    let json;
    try { json = JSON.parse(dataStr); } catch { return; }
    const choice = json.choices?.[0];
    if (!choice) return;
    const delta = choice.delta || {};

    if (delta.content) {
      content += delta.content;
      handlers.onContentDelta?.(delta.content);
    }

    if (delta.tool_calls) {
      for (const tc of delta.tool_calls) {
        const idx = tc.index ?? 0;
        if (!toolCallsMap.has(idx)) {
          toolCallsMap.set(idx, { id: tc.id || "", name: "", argsStr: "" });
        }
        const entry = toolCallsMap.get(idx);
        if (tc.id) entry.id = tc.id;
        if (tc.function?.name) entry.name += tc.function.name;
        if (tc.function?.arguments) entry.argsStr += tc.function.arguments;
      }
    }

    if (choice.finish_reason) finishReason = choice.finish_reason;
  };

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    let nl;
    while ((nl = buffer.indexOf("\n")) >= 0) {
      let line = buffer.slice(0, nl).trim();
      buffer = buffer.slice(nl + 1);
      if (!line || line.startsWith(":")) continue;
      if (line.startsWith("data:")) {
        const dataStr = line.slice(5).trim();
        handleEvent(dataStr);
      }
    }
  }

  const toolCalls = [...toolCallsMap.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([, v]) => ({
      id: v.id || `call_${Math.random().toString(36).slice(2, 10)}`,
      type: "function",
      function: { name: v.name, arguments: v.argsStr || "{}" },
    }));

  return { content, toolCalls, finishReason };
}

export async function testConfig(config) {
  const base = normalizeBaseUrl(config.baseUrl);
  const endpoint = `${base}/chat/completions`;
  const resp = await fetch(endpoint, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${config.apiKey || ""}`,
    },
    body: JSON.stringify({
      model: config.model,
      messages: [{ role: "user", content: "ping，请只回复 pong" }],
      stream: false,
      max_tokens: 16,
    }),
  });
  const text = await resp.text();
  if (!resp.ok) {
    throw new Error(`${resp.status}: ${text.slice(0, 300)}`);
  }
  let reply = "";
  try {
    reply = JSON.parse(text).choices?.[0]?.message?.content || "";
  } catch { /* ignore */ }
  return reply;
}

File: public/demos/topic_146267/VoxHire/backend/llm.py
from __future__ import annotations

from urllib.parse import urlparse

import httpx

from .models import LlmConnectionRequest, LlmConnectionResult


def chat_completion_url(base_url: str) -> str:
    normalized = base_url.rstrip("/")
    if normalized.endswith("/chat/completions"):
        return normalized
    if not normalized.endswith("/v1"):
        normalized = f"{normalized}/v1"
    return f"{normalized}/chat/completions"


async def test_openai_compatible_connection(config: LlmConnectionRequest) -> LlmConnectionResult:
    parsed = urlparse(config.base_url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        return LlmConnectionResult(ok=False, message="请输入完整的 HTTP(S) API 地址。", model=config.model)

    try:
        async with httpx.AsyncClient(timeout=20.0) as client:
            response = await client.post(
                chat_completion_url(config.base_url),
                headers={"Authorization": f"Bearer {config.api_key}"},
                json={
                    "model": config.model,
                    "messages": [{"role": "user", "content": "仅回复 OK"}],
                    "max_tokens": 4,
                    "temperature": 0,
                },
            )
        if response.is_success:
            return LlmConnectionResult(ok=True, message="连接成功，可使用该模型。", model=config.model)
        detail = response.text[:180].replace("\n", " ")
        return LlmConnectionResult(ok=False, message=f"请求失败（HTTP {response.status_code}）：{detail}", model=config.model)
    except httpx.TimeoutException:
        return LlmConnectionResult(ok=False, message="连接超时，请检查地址、网络或模型服务状态。", model=config.model)
    except httpx.HTTPError as exc:
        return LlmConnectionResult(ok=False, message=f"无法连接 API：{str(exc)[:160]}", model=config.model)
